File: Controllers/NativeExpressAdVC.m
//
//  FeedVC.m
//  Demo
//
//  Created by mac on 2020/8/13.
//  Copyright © 2020 MobiusAd. All rights reserved.
//

#import "NativeExpressAdVC.h"
#import <AdKleinSDK/AdKleinSDKNativeExpressAd.h>

@interface NativeExpressAdVC () <AdKleinSDKNativeExpressAdDelegate, UITableViewDelegate, UITableViewDataSource>

@property(nonatomic, strong) AdKleinSDKNativeExpressAd *adLoader;
@property(nonatomic, strong) UIView *showAdView;

@property (nonatomic, strong) NSMutableArray *dataArray;
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, assign) CGSize adSize;

@end

@implementation NativeExpressAdVC

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    self.dataArray = [NSMutableArray new];

    self.showBtn.hidden = YES;
    CGSize size = self.view.frame.size;
    self.tableView = [[UITableView alloc] initWithFrame:CGRectMake(0, 100, size.width, size.height-100)];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"UITableViewCell"];
    self.tableView.backgroundColor = [UIColor clearColor];
    [self.view addSubview:self.tableView];
    [self.view sendSubviewToBack:self.tableView];
}

- (void)loadClick {
    if (!self.adLoader) {
        self.adSize = CGSizeMake(self.view.frame.size.width, 0);

        self.adLoader = [[AdKleinSDKNativeExpressAd alloc] initWithPlacementId:CONST_FEED_EXPRESS_ID viewController:self];
        self.adLoader.delegate = self;
        self.adLoader.adSize = self.adSize;
        self.adLoader.videoMuted = NO;
        self.adLoader.adCount = 3;
    }

    [self.adLoader load];
}

- (void)ak_nativeExpressAdDidLoad:(AdKleinSDKNativeExpressAd *)nativeAd withAdViews:(NSArray<__kindof UIView<AdKleinSDKNativeExpressAdViewDelegate> *> *)adViews {
    // 将获取到的广告视图进行模板渲染
    [adViews enumerateObjectsUsingBlock:^(UIView<AdKleinSDKNativeExpressAdViewDelegate> *adView, NSUInteger idx, BOOL * _Nonnull stop) {
        [adView render];
    }];

    NSString *func = [NSString stringWithFormat:@"%s",__func__];
    [self showString:func];
}

- (void)ak_nativeExpressAdDidFail:(AdKleinSDKNativeExpressAd *)nativeAd withError:(NSError *)error {
    [self showError:error];
}
- (void)ak_nativeExpressAdDidRenderSuccess:(AdKleinSDKNativeExpressAd *)nativeAd adView:(__kindof UIView<AdKleinSDKNativeExpressAdViewDelegate> *)adView {
    // 渲染成功才能拿到真实的高度
    dispatch_async(dispatch_get_main_queue(), ^{
        for (int i = 0; i < 6; i ++) {
            [self.dataArray addObject:[NSNull null]];
        }
        [self.dataArray addObject:adView];
        [self.tableView reloadData];
    });

    NSString *func = [NSString stringWithFormat:@"%s",__func__];
    [self showString:func];
}
- (void)ak_nativeExpressAdDidRenderFail:(AdKleinSDKNativeExpressAd *)nativeAd adView:(__kindof UIView<AdKleinSDKNativeExpressAdViewDelegate> *)adView {
    NSString *func = [NSString stringWithFormat:@"%s",__func__];
    [self showString:func];
}
- (void)ak_nativeExpressAdDidShow:(AdKleinSDKNativeExpressAd *)nativeAd adView:(__kindof UIView<AdKleinSDKNativeExpressAdViewDelegate> *)adView {
    NSString *func = [NSString stringWithFormat:@"%s",__func__];
    [self showString:func];
}
- (void)ak_nativeExpressAdDidClick:(AdKleinSDKNativeExpressAd *)nativeAd adView:(__kindof UIView<AdKleinSDKNativeExpressAdViewDelegate> *)adView {
    NSString *func = [NSString stringWithFormat:@"%s",__func__];
    [self showString:func];
}
- (void)ak_nativeExpressAdDidClose:(AdKleinSDKNativeExpressAd *)nativeAd adView:(__kindof UIView<AdKleinSDKNativeExpressAdViewDelegate> *)adView {
    [adView removeFromSuperview];
    [self.dataArray removeObject:adView];
    [self.tableView reloadData];

    NSString *func = [NSString stringWithFormat:@"%s",__func__];
    [self showString:func];
}

#pragma mark - tableView
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"UITableViewCell"];

    cell.backgroundColor = [UIColor clearColor];
    cell.textLabel.text = @"";

    if ([self.dataArray[indexPath.row] conformsToProtocol:@protocol(AdKleinSDKNativeExpressAdViewDelegate)]) {
        UIView<AdKleinSDKNativeExpressAdViewDelegate> *view = (UIView<AdKleinSDKNativeExpressAdViewDelegate> *)self.dataArray[indexPath.row];
        [cell.contentView addSubview:view];
    } else {
        cell.textLabel.text = [NSString stringWithFormat:@"%ld",indexPath.row];
        cell.backgroundColor = [UIColor grayColor];
    }

    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if ([self.dataArray[indexPath.row] conformsToProtocol:@protocol(AdKleinSDKNativeExpressAdViewDelegate)]) {
        UIView<AdKleinSDKNativeExpressAdViewDelegate> *view = (UIView<AdKleinSDKNativeExpressAdViewDelegate> *)self.dataArray[indexPath.row];
        NSLog(@"rowHeight = %f", view.frame.size.height);
        return view.frame.size.height;
    } else {
        return 50;
    }
}


@end

File: SwiftDemo/Controllers/NativeExpressAdVC.swift
//
//  NativeExpressAdVC.swift
//  SwiftDemo
//
//  Created by mirari on 2021/9/16.
//  Copyright © 2021 MobiusAd. All rights reserved.
//

import SwiftUI

class NativeExpressAdVC: BaseViewController, AdKleinSDKNativeExpressAdDelegate, UITableViewDelegate, UITableViewDataSource {
    var adLoader: AdKleinSDKNativeExpressAd?
    var dataArray:Array = Array<Any>.init()
    var tableView: UITableView?
    var adSize = CGSize.zero
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        showBtn.isHidden = true;
        
        let size = view.frame.size
        tableView = UITableView(frame: CGRect(x: 0, y: 100, width: size.width, height: size.height - 100))
        tableView?.delegate = self
        tableView?.dataSource = self
        tableView?.register(UITableViewCell.self, forCellReuseIdentifier: "UITableViewCell")
        tableView?.backgroundColor = UIColor.clear
        view.addSubview(tableView!)
        view.sendSubviewToBack(tableView!)
    }

    override func loadClick() {
        if (adLoader == nil) {
            adSize = CGSize(width: view.frame.size.width, height: 0)

            adLoader = AdKleinSDKNativeExpressAd(placementId: Constant.NATIVE_EXPRESS_ID, viewController: self)
            adLoader?.delegate = self
            adLoader?.adSize = adSize
            adLoader?.videoMuted = false
            adLoader?.adCount = 3
        }

        adLoader?.load()
    }
        
    func ak_nativeExpressAdDidLoad(_ nativeAd: AdKleinSDKNativeExpressAd, withAdViews adViews: [UIView & AdKleinSDKNativeExpressAdViewDelegate]){
        // 将获取到的广告视图进行模板渲染
        for adView in adViews {
            adView.render()
        }
        showString(#function)
    }

    func ak_nativeExpressAdDidFail(_ nativeAd: AdKleinSDKNativeExpressAd, withError error: Error) {
        showError(error)
    }
    
    func ak_nativeExpressAdDidRenderSuccess(_ nativeAd: AdKleinSDKNativeExpressAd, adView: UIView & AdKleinSDKNativeExpressAdViewDelegate){
        // 渲染成功才能拿到真实的高度
        DispatchQueue.main.async(execute: { [self] in
            for _ in 0..<6 {
                dataArray.append(NSNull())
            }
            dataArray.append(adView)
            tableView?.reloadData()
        })
        showString(#function)
    }
    
    func ak_nativeExpressAdDidRenderFail(_ nativeAd: AdKleinSDKNativeExpressAd, adView: UIView & AdKleinSDKNativeExpressAdViewDelegate){
        showString(#function)
    }
    
    func ak_nativeExpressAdDidShow(_ nativeAd: AdKleinSDKNativeExpressAd, adView: UIView & AdKleinSDKNativeExpressAdViewDelegate){
        showString(#function)
    }
    
    func ak_nativeExpressAdDidClick(_ nativeAd: AdKleinSDKNativeExpressAd, adView: UIView & AdKleinSDKNativeExpressAdViewDelegate){
        showString(#function)
    }
    
    func ak_nativeExpressAdDidClose(_ nativeAd: AdKleinSDKNativeExpressAd, adView: UIView & AdKleinSDKNativeExpressAdViewDelegate){
        showString(#function)
    }
    
    // MARK: - tableView
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell(style: .default, reuseIdentifier: "UITableViewCell")

        cell.backgroundColor = UIColor.clear
        cell.textLabel?.text = ""
        
        if dataArray[indexPath.row] is AdKleinSDKNativeExpressAdViewDelegate {
            let view = dataArray[indexPath.row] as! UIView
            cell.contentView.addSubview(view)
        } else {
            cell.textLabel?.text = String(format: "%ld", indexPath.row)
            cell.backgroundColor = UIColor.gray
        }
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if dataArray[indexPath.row] is AdKleinSDKNativeExpressAdViewDelegate {
            let view = dataArray[indexPath.row] as! UIView
            let height = view.frame.size.height
            print("rowHeight = \(height)")
            return height
        } else {
            return 50
        }
    }
}

